File: 05_arrays/05_17_buy_and_sell_stock/buy_and_sell_stock.js
const maxProfit = (arr = []) => {
    if (arr.length === 0) return 0;
    let profit = 0;
    let min = arr[0];

    for (let i = 1; i < arr.length; i++) {
        const current_profit = arr[i] - min;
        profit = Math.max(profit, current_profit);
        min = Math.min(min, arr[i]);
    }

    return profit;
}

let arr = [7, 1, 5, 3, 6, 4];
console.log("Values:");
console.log(arr.join(" "));
let result = maxProfit(arr);
console.log("The maximum profit is:", result);

File: 05_arrays/05_17_buy_and_sell_stock/BuyAndSellStock.java
import java.util.Arrays;

public class BuyAndSellStock {
    public static void main(String[] args) {
        int[] arr = {7, 1, 5, 3, 6, 4};
        System.out.println("Values:");
        System.out.println(Arrays.toString(arr));
        int result = maxProfit(arr);
        System.out.println("The maximum profit is: " + result);
    }

    private static int maxProfit(int[] arr) {
        if (arr.length == 0) {
            return 0;
        }

        int profit = 0;
        int min = arr[0];

        for (int i = 1; i < arr.length; i++) {
            int currentProfit = arr[i] - min;
            profit = Math.max(profit, currentProfit);
            min = Math.min(min, arr[i]);
        }

        return profit;
    }
}
